File: android/src/main/java/com/github/quadflask/react/navermap/RNNaverMapViewManager.java
package com.github.quadflask.react.navermap;

import android.graphics.Rect;
import android.view.View;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.airbnb.android.react.maps.SizeReportingShadowNode;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.common.MapBuilder;
import com.facebook.react.uimanager.LayoutShadowNode;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.ViewGroupManager;
import com.facebook.react.uimanager.annotations.ReactProp;
import com.naver.maps.geometry.LatLng;
import com.naver.maps.geometry.LatLngBounds;
import com.naver.maps.map.NaverMap;
import com.naver.maps.map.NaverMapOptions;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static com.github.quadflask.react.navermap.ReactUtil.getDoubleOrNull;
import static com.github.quadflask.react.navermap.ReactUtil.getInt;
import static com.github.quadflask.react.navermap.ReactUtil.toNaverLatLng;

public class RNNaverMapViewManager extends ViewGroupManager<RNNaverMapViewContainer> {
    private final ReactApplicationContext appContext;

    private static final int ANIMATE_TO_REGION = 1;
    private static final int ANIMATE_TO_COORDINATE = 2;
    private static final int ANIMATE_TO_TWO_COORDINATES = 3;
    private static final int SET_LOCATION_TRACKING_MODE = 4;
    private static final int ANIMATE_TO_COORDINATES = 6;
    private static final int SET_LAYER_GROUP_ENABLED = 7;
    private static final int LOCATION_PERMISSION_REQUEST_CODE = 100;
    private static final List<String> LAYER_GROUPS = Collections.unmodifiableList(Arrays.asList(
            NaverMap.LAYER_GROUP_BUILDING,
            NaverMap.LAYER_GROUP_TRANSIT,
            NaverMap.LAYER_GROUP_BICYCLE,
            NaverMap.LAYER_GROUP_TRAFFIC,
            NaverMap.LAYER_GROUP_CADASTRAL,
            NaverMap.LAYER_GROUP_MOUNTAIN
    ));

    public RNNaverMapViewManager(ReactApplicationContext context) {
        super();
        this.appContext = context;
    }

    @NonNull
    @Override
    public String getName() {
        return "RNNaverMapView";
    }

    @NonNull
    @Override
    protected RNNaverMapViewContainer createViewInstance(@NonNull ThemedReactContext reactContext) {
        return new RNNaverMapViewContainer(reactContext, appContext, getNaverMapViewOptions());
    }

    @Override
    public void onDropViewInstance(@NonNull RNNaverMapViewContainer view) {
        super.onDropViewInstance(view);
        view.onDropViewInstance();
    }

    protected NaverMapOptions getNaverMapViewOptions() {
        return null;
    }

    @ReactProp(name = "center")
    public void setCenter(RNNaverMapViewContainer mapView, @Nullable ReadableMap option) {
        if (option != null) {
            mapView.setCenter(
                    toNaverLatLng(option),
                    getDoubleOrNull(option, "zoom"),
                    getDoubleOrNull(option, "tilt"),
                    getDoubleOrNull(option, "bearing"));
        }
    }

    @ReactProp(name = "showsMyLocationButton", defaultBoolean = false)
    public void showsMyLocationButton(RNNaverMapViewContainer mapView, boolean show) {
        mapView.showsMyLocationButton(show);
    }

    @ReactProp(name = "compass", defaultBoolean = false)
    public void setCompassEnabled(RNNaverMapViewContainer mapView, boolean show) {
        mapView.setCompassEnabled(show);
    }

    @ReactProp(name = "scaleBar", defaultBoolean = false)
    public void setScaleBarEnabled(RNNaverMapViewContainer mapView, boolean show) {
        mapView.setScaleBarEnabled(show);
    }

    @ReactProp(name = "zoomControl", defaultBoolean = false)
    public void setZoomControlEnabled(RNNaverMapViewContainer mapView, boolean show) {
        mapView.setZoomControlEnabled(show);
    }

    @ReactProp(name = "mapType", defaultInt = 0)
    public void setMapType(RNNaverMapViewContainer mapView, int mapType) {
        mapView.setMapType(NaverMap.MapType.values()[mapType]);
    }

    @ReactProp(name = "minZoomLevel", defaultFloat = 0)
    public void setMinZoom(RNNaverMapViewContainer mapView, float minZoomLevel) {
        mapView.setMinZoom(minZoomLevel);
    }

    @ReactProp(name = "maxZoomLevel", defaultFloat = 21.0f)
    public void setMaxZoom(RNNaverMapViewContainer mapView, float maxZoomLevel) {
        mapView.setMaxZoom(maxZoomLevel);
    }

    @ReactProp(name = "buildingHeight", defaultFloat = 1.0f)
    public void setBuildingHeight(RNNaverMapViewContainer mapView, float height) {
        mapView.setBuildingHeight(height);
    }

    @ReactProp(name = "locationTrackingMode", defaultInt = 0)
    public void locationTrackingMode(RNNaverMapViewContainer mapView, int mode) {
        if (mode >= 0)
            mapView.setLocationTrackingMode(mode);
    }

    @ReactProp(name = "tilt", defaultInt = 0)
    public void setTilt(RNNaverMapViewContainer mapView, int tilt) {
        mapView.setTilt(tilt);
    }

    @ReactProp(name = "bearing", defaultInt = 0)
    public void setBearing(RNNaverMapViewContainer mapView, int bearing) {
        mapView.setBearing(bearing);
    }

    @ReactProp(name = "layerGroup")
    public void setLayerGroupEnabled(RNNaverMapViewContainer mapView, @Nullable ReadableMap option) {
        if (option != null) {
            for (String layerGroup : LAYER_GROUPS) {
                mapView.setLayerGroupEnabled(layerGroup, option.getBoolean(layerGroup));
            }
        } else {
            for (String layerGroup : LAYER_GROUPS) {
                mapView.setLayerGroupEnabled(layerGroup, false);
            }
        }
    }

    @ReactProp(name = "nightMode", defaultBoolean = false)
    public void setNightModeEnabled(RNNaverMapViewContainer mapView, boolean enable) {
        mapView.setNightModeEnabled(enable);
    }

    @ReactProp(name = "mapPadding")
    public void setMapPadding(RNNaverMapViewContainer mapView, @Nullable ReadableMap padding) {
        final Rect rect = getRect(padding, mapView.getResources().getDisplayMetrics().density);

        mapView.setMapPadding(rect.left, rect.top, rect.right, rect.bottom);
    }

    @ReactProp(name = "logoMargin")
    public void setLogoMargin(RNNaverMapViewContainer mapView, @Nullable ReadableMap margin) {
        final Rect rect = getRect(margin, mapView.getResources().getDisplayMetrics().density);

        mapView.setLogoMargin(rect.left, rect.top, rect.right, rect.bottom);
    }

    @ReactProp(name = "logoGravity")
    public void setLogoGravity(RNNaverMapViewContainer mapView, int gravity) {
        mapView.setLogoGravity(gravity);
    }

    @ReactProp(name = "scrollGesturesEnabled", defaultBoolean = true)
    public void setScrollGesturesEnabled(RNNaverMapViewContainer mapView, boolean enabled) {
        mapView.setScrollGesturesEnabled(enabled);
    }

    @ReactProp(name = "zoomGesturesEnabled", defaultBoolean = true)
    public void setZoomGesturesEnabled(RNNaverMapViewContainer mapView, boolean enabled) {
        mapView.setZoomGesturesEnabled(enabled);
    }

    @ReactProp(name = "tiltGesturesEnabled", defaultBoolean = true)
    public void setTiltGesturesEnabled(RNNaverMapViewContainer mapView, boolean enabled) {
        mapView.setTiltGesturesEnabled(enabled);
    }

    @ReactProp(name = "rotateGesturesEnabled", defaultBoolean = true)
    public void setRotateGesturesEnabled(RNNaverMapViewContainer mapView, boolean enabled) {
        mapView.setRotateGesturesEnabled(enabled);
    }

    @ReactProp(name = "stopGesturesEnabled", defaultBoolean = true)
    public void setStopGesturesEnabled(RNNaverMapViewContainer mapView, boolean enabled) {
        mapView.setStopGesturesEnabled(enabled);
    }

    @ReactProp(name = "liteModeEnabled", defaultBoolean = false)
    public void setLiteModeEnabled(RNNaverMapViewContainer mapView, boolean enabled) {
        mapView.setLiteModeEnabled(enabled);
    }

    private Rect getRect(@Nullable ReadableMap padding, float density) {
        int left = 0;
        int top = 0;
        int right = 0;
        int bottom = 0;

        if (padding != null) {
            if (padding.hasKey("left")) {
                left = Math.round(((float) padding.getDouble("left") * density));
            }
            if (padding.hasKey("top")) {
                top = Math.round(((float) padding.getDouble("top") * density));
            }
            if (padding.hasKey("right")) {
                right = Math.round(((float) padding.getDouble("right") * density));
            }
            if (padding.hasKey("bottom")) {
                bottom = Math.round(((float) padding.getDouble("bottom") * density));
            }
        }

        return new Rect(left, top, right, bottom);
    }

    @Override
    public void receiveCommand(@NonNull RNNaverMapViewContainer mapView, int commandId, @Nullable ReadableArray args) {
        switch (commandId) {
            case ANIMATE_TO_REGION: {
                final ReadableMap region = args.getMap(0);
                final double lat = region.getDouble("latitude");
                final double lng = region.getDouble("longitude");
                final double latDelta = region.getDouble("latitudeDelta");
                final double lngDelta = region.getDouble("longitudeDelta");

                LatLngBounds bounds = new LatLngBounds(
                        new LatLng(lat - latDelta / 2, lng - lngDelta / 2), // southwest
                        new LatLng(lat + latDelta / 2, lng + lngDelta / 2)  // northeast
                );
                mapView.moveCameraFitBound(bounds, 0, 0, 0, 0);

                break;
            }
            case ANIMATE_TO_TWO_COORDINATES: {
                float density = mapView.getResources().getDisplayMetrics().density;

                final ReadableMap region1 = args.getMap(0);
                final ReadableMap region2 = args.getMap(1);
                final int padding = args.size() == 3 ? args.getInt(2) : 56;

                final double lat1 = region1.getDouble("latitude");
                final double lng1 = region1.getDouble("longitude");
                final double lat2 = region2.getDouble("latitude");
                final double lng2 = region2.getDouble("longitude");

                mapView.zoomTo(new LatLngBounds(
                                new LatLng(Math.max(lat1, lat2), Math.min(lng1, lng2)),
                                new LatLng(Math.min(lat1, lat2), Math.max(lng1, lng2))),
                        Math.round(padding * density));

                break;
            }
            case ANIMATE_TO_COORDINATES:
                ReadableArray coordinatesArray = args.getArray(0);
                ReadableMap edgePadding = args.getMap(1);

                LatLngBounds.Builder builder = new LatLngBounds.Builder();
                for (int i = 0; i < coordinatesArray.size(); i++) {
                    builder.include(toNaverLatLng(coordinatesArray.getMap(i)));
                }

                int left = getInt(edgePadding, "left", 0);
                int top = getInt(edgePadding, "top", 0);
                int right = getInt(edgePadding, "right", 0);
                int bottom = getInt(edgePadding, "bottom", 0);

                mapView.moveCameraFitBound(builder.build(), left, top, right, bottom);

                break;
            case ANIMATE_TO_COORDINATE:
                mapView.setCenter(toNaverLatLng(args.getMap(0)));
                break;

            case SET_LOCATION_TRACKING_MODE:
                mapView.setLocationTrackingMode(args.getInt(0));
                break;

            case SET_LAYER_GROUP_ENABLED:
                String group = args.getString(0);
                Boolean enable = args.getBoolean(1);

                switch (group) {
                    case "building":
                        mapView.setLayerGroupEnabled(NaverMap.LAYER_GROUP_BUILDING, enable);
                        break;
                    case "transit":
                        mapView.setLayerGroupEnabled(NaverMap.LAYER_GROUP_TRANSIT, enable);
                        break;
                    case "bike":
                        mapView.setLayerGroupEnabled(NaverMap.LAYER_GROUP_BICYCLE, enable);
                        break;
                    case "ctt":
                        mapView.setLayerGroupEnabled(NaverMap.LAYER_GROUP_TRAFFIC, enable);
                        break;
                    case "landparcel":
                        mapView.setLayerGroupEnabled(NaverMap.LAYER_GROUP_CADASTRAL, enable);
                        break;
                    case "mountain":
                        mapView.setLayerGroupEnabled(NaverMap.LAYER_GROUP_MOUNTAIN, enable);
                        break;
                }
                break;
        }
    }

    @Override
    public java.util.Map<String, Integer> getCommandsMap() {
        return MapBuilder.<String, Integer>builder()
                .put("animateToRegion", ANIMATE_TO_REGION)
                .put("animateToCoordinate", ANIMATE_TO_COORDINATE)
                .put("animateToTwoCoordinates", ANIMATE_TO_TWO_COORDINATES)
                .put("setLocationTrackingMode", SET_LOCATION_TRACKING_MODE)
                .put("animateToCoordinates", ANIMATE_TO_COORDINATES)
                .put("setLayerGroupEnabled", SET_LAYER_GROUP_ENABLED)
                .build();
    }

    @Override
    public java.util.Map<String, Object> getExportedCustomBubblingEventTypeConstants() {
        final MapBuilder.Builder<String, Object> builder = MapBuilder.builder();
        for (String eventName : RNNaverMapViewProps.EVENT_NAMES)
            builder.put(eventName, bubbled(eventName));
        return builder.build();
    }

    @Override
    public LayoutShadowNode createShadowNodeInstance() {
        // A custom shadow node is needed in order to pass back the width/height of the map to the
        // view manager so that it can start applying camera moves with bounds.
        return new SizeReportingShadowNode();
    }

    @Override
    public void addView(RNNaverMapViewContainer parent, View child, int index) {
        parent.addFeature(child, index);
    }

    @Override
    public int getChildCount(RNNaverMapViewContainer view) {
        return view.getFeatureCount();
    }

    @Override
    public View getChildAt(RNNaverMapViewContainer view, int index) {
        return view.getFeatureAt(index);
    }

    @Override
    public void removeViewAt(RNNaverMapViewContainer parent, int index) {
        parent.removeFeatureAt(index);
    }

    @Override
    public boolean needsCustomLayoutForChildren() {
        return true;
    }

    private java.util.Map<String, Object> bubbled(String callbackName) {
        return MapBuilder.of("phasedRegistrationNames", MapBuilder.of("bubbled", callbackName));
    }
}

File: android/src/main/java/com/github/quadflask/react/navermap/RNNaverMapViewManager.kt
package com.github.quadflask.react.navermap

import android.graphics.Rect
import android.util.Log
import android.view.View
import com.airbnb.android.react.maps2.SizeReportingShadowNode
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.uimanager.LayoutShadowNode
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.ViewGroupManager
import com.facebook.react.uimanager.annotations.ReactProp
import com.github.quadflask.react.navermap.ReactUtil.getDoubleOrNull
import com.github.quadflask.react.navermap.ReactUtil.getInt
import com.github.quadflask.react.navermap.ReactUtil.toLatLngBounds
import com.github.quadflask.react.navermap.ReactUtil.toNaverLatLng
import com.naver.maps.geometry.LatLng
import com.naver.maps.geometry.LatLngBounds
import com.naver.maps.map.NaverMap
import com.naver.maps.map.NaverMapOptions
import com.naver.maps.map.util.FusedLocationSource
import java.util.*
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

open class RNNaverMapViewManager(private val appContext: ReactApplicationContext) : ViewGroupManager<RNNaverMapViewContainer>() {
  private val locationSource: FusedLocationSource = FusedLocationSource(appContext.currentActivity!!, 0x1000)

  companion object {
    private const val ANIMATE_TO_REGION = 1
    private const val ANIMATE_TO_COORDINATE = 2
    private const val ANIMATE_TO_TWO_COORDINATES = 3
    private const val SET_LOCATION_TRACKING_MODE = 4
    private const val ANIMATE_TO_COORDINATES = 6
    private const val SET_LAYER_GROUP_ENABLED = 7
    private val LAYER_GROUPS = Collections.unmodifiableList(
      listOf(
        NaverMap.LAYER_GROUP_BUILDING,
        NaverMap.LAYER_GROUP_TRANSIT,
        NaverMap.LAYER_GROUP_BICYCLE,
        NaverMap.LAYER_GROUP_TRAFFIC,
        NaverMap.LAYER_GROUP_CADASTRAL,
        NaverMap.LAYER_GROUP_MOUNTAIN
      )
    )
  }

  override fun getName() = "RNNaverMapView"

  override fun createViewInstance(reactContext: ThemedReactContext): RNNaverMapViewContainer {
    Log.e("RNN", "RNNaverMapViewManager.createViewInstance")
    return RNNaverMapViewContainer(reactContext, appContext, locationSource, naverMapViewOptions)
  }

  override fun onDropViewInstance(view: RNNaverMapViewContainer) {
    super.onDropViewInstance(view)
    view.onDropViewInstance()
  }

  protected open val naverMapViewOptions: NaverMapOptions?
    get() = NaverMapOptions()

  @ReactProp(name = "center")
  fun setCenter(mapView: RNNaverMapViewContainer, option: ReadableMap?) {
    if (option != null) {
      mapView.setCenter(
        toNaverLatLng(option),
        getDoubleOrNull(option, "zoom"),
        getDoubleOrNull(option, "tilt"),
        getDoubleOrNull(option, "bearing")
      )
    }
  }

  @ReactProp(name = "showsMyLocationButton", defaultBoolean = false)
  fun showsMyLocationButton(mapView: RNNaverMapViewContainer, show: Boolean) {
    mapView.showsMyLocationButton(show)
  }

  @ReactProp(name = "compass", defaultBoolean = false)
  fun setCompassEnabled(mapView: RNNaverMapViewContainer, show: Boolean) {
    mapView.setCompassEnabled(show)
  }

  @ReactProp(name = "scaleBar", defaultBoolean = false)
  fun setScaleBarEnabled(mapView: RNNaverMapViewContainer, show: Boolean) {
    mapView.setScaleBarEnabled(show)
  }

  @ReactProp(name = "zoomControl", defaultBoolean = false)
  fun setZoomControlEnabled(mapView: RNNaverMapViewContainer, show: Boolean) {
    mapView.setZoomControlEnabled(show)
  }

  @ReactProp(name = "mapType", defaultInt = 0)
  fun setMapType(mapView: RNNaverMapViewContainer, mapType: Int) {
    mapView.setMapType(NaverMap.MapType.values()[mapType])
  }

  @ReactProp(name = "minZoomLevel", defaultFloat = 0f)
  fun setMinZoom(mapView: RNNaverMapViewContainer, minZoomLevel: Float) {
    mapView.setMinZoom(minZoomLevel)
  }

  @ReactProp(name = "maxZoomLevel", defaultFloat = 21.0f)
  fun setMaxZoom(mapView: RNNaverMapViewContainer, maxZoomLevel: Float) {
    mapView.setMaxZoom(maxZoomLevel)
  }

  @ReactProp(name = "extent")
  fun setExtent(mapView: RNNaverMapViewContainer, option: ReadableArray?) {
    mapView.setExtent(option?.let { toLatLngBounds(it.getMap(0), it.getMap(1)) })
  }

  @ReactProp(name = "buildingHeight", defaultFloat = 1.0f)
  fun setBuildingHeight(mapView: RNNaverMapViewContainer, height: Float) {
    mapView.setBuildingHeight(height)
  }

  @ReactProp(name = "locationTrackingMode", defaultInt = 0)
  fun locationTrackingMode(mapView: RNNaverMapViewContainer, mode: Int) {
    if (mode >= 0) mapView.setLocationTrackingMode(mode)
  }

  @ReactProp(name = "tilt", defaultInt = 0)
  fun setTilt(mapView: RNNaverMapViewContainer, tilt: Int) {
    mapView.setTilt(tilt)
  }

  @ReactProp(name = "bearing", defaultInt = 0)
  fun setBearing(mapView: RNNaverMapViewContainer, bearing: Int) {
    mapView.setBearing(bearing)
  }

  @ReactProp(name = "layerGroup")
  fun setLayerGroupEnabled(mapView: RNNaverMapViewContainer, option: ReadableMap?) {
    for (layerGroup in LAYER_GROUPS)
      mapView.setLayerGroupEnabled(layerGroup, option?.getBoolean(layerGroup!!) ?: false)
  }

  @ReactProp(name = "nightMode", defaultBoolean = false)
  fun setNightModeEnabled(mapView: RNNaverMapViewContainer, enable: Boolean) {
    mapView.setNightModeEnabled(enable)
  }

  @ReactProp(name = "mapPadding")
  fun setMapPadding(mapView: RNNaverMapViewContainer, padding: ReadableMap?) {
    val rect = getRect(padding, mapView.resources.displayMetrics.density)
    mapView.setMapPadding(rect.left, rect.top, rect.right, rect.bottom)
  }

  @ReactProp(name = "logoMargin")
  fun setLogoMargin(mapView: RNNaverMapViewContainer, margin: ReadableMap?) {
    val rect = getRect(margin, mapView.resources.displayMetrics.density)
    mapView.setLogoMargin(rect.left, rect.top, rect.right, rect.bottom)
  }

  @ReactProp(name = "logoGravity")
  fun setLogoGravity(mapView: RNNaverMapViewContainer, gravity: Int) {
    mapView.setLogoGravity(gravity)
  }

  @ReactProp(name = "scrollGesturesEnabled", defaultBoolean = true)
  fun setScrollGesturesEnabled(mapView: RNNaverMapViewContainer, enabled: Boolean) {
    mapView.setScrollGesturesEnabled(enabled)
  }

  @ReactProp(name = "zoomGesturesEnabled", defaultBoolean = true)
  fun setZoomGesturesEnabled(mapView: RNNaverMapViewContainer, enabled: Boolean) {
    mapView.setZoomGesturesEnabled(enabled)
  }

  @ReactProp(name = "tiltGesturesEnabled", defaultBoolean = true)
  fun setTiltGesturesEnabled(mapView: RNNaverMapViewContainer, enabled: Boolean) {
    mapView.setTiltGesturesEnabled(enabled)
  }

  @ReactProp(name = "rotateGesturesEnabled", defaultBoolean = true)
  fun setRotateGesturesEnabled(mapView: RNNaverMapViewContainer, enabled: Boolean) {
    mapView.setRotateGesturesEnabled(enabled)
  }

  @ReactProp(name = "stopGesturesEnabled", defaultBoolean = true)
  fun setStopGesturesEnabled(mapView: RNNaverMapViewContainer, enabled: Boolean) {
    mapView.setStopGesturesEnabled(enabled)
  }

  @ReactProp(name = "liteModeEnabled", defaultBoolean = false)
  fun setLiteModeEnabled(mapView: RNNaverMapViewContainer, enabled: Boolean) {
    mapView.setLiteModeEnabled(enabled)
  }

  private fun getRect(padding: ReadableMap?, density: Float): Rect {
    return Rect(0, 0, 0, 0).apply {
      if (padding != null) {
        if (padding.hasKey("left"))
          left = (padding.getDouble("left").toFloat() * density).roundToInt()
        if (padding.hasKey("top"))
          top = (padding.getDouble("top").toFloat() * density).roundToInt()
        if (padding.hasKey("right"))
          right = (padding.getDouble("right").toFloat() * density).roundToInt()
        if (padding.hasKey("bottom"))
          bottom = (padding.getDouble("bottom").toFloat() * density).roundToInt()
      }
    }
  }

  override fun receiveCommand(mapView: RNNaverMapViewContainer, commandId: Int, args: ReadableArray?) {
    if (args == null) return

    when (commandId) {
      ANIMATE_TO_REGION -> {
        val region = args.getMap(0)
        val lat = region.getDouble("latitude")
        val lng = region.getDouble("longitude")
        val latDelta = region.getDouble("latitudeDelta")
        val lngDelta = region.getDouble("longitudeDelta")
        val bounds = LatLngBounds(
          LatLng(lat - latDelta / 2, lng - lngDelta / 2),  // southwest
          LatLng(lat + latDelta / 2, lng + lngDelta / 2) // northeast
        )
        mapView.moveCameraFitBound(bounds, 0, 0, 0, 0)
      }
      ANIMATE_TO_TWO_COORDINATES -> {
        val density = mapView.resources.displayMetrics.density
        val region1 = args.getMap(0)
        val region2 = args.getMap(1)
        val padding = if (args.size() == 3) args.getInt(2) else 56
        val lat1 = region1.getDouble("latitude")
        val lng1 = region1.getDouble("longitude")
        val lat2 = region2.getDouble("latitude")
        val lng2 = region2.getDouble("longitude")
        mapView.zoomTo(
          LatLngBounds(
            LatLng(max(lat1, lat2), min(lng1, lng2)),
            LatLng(min(lat1, lat2), max(lng1, lng2))
          ),
          (padding * density).roundToInt()
        )
      }
      ANIMATE_TO_COORDINATES -> {
        val coordinatesArray = args.getArray(0)
        val edgePadding = args.getMap(1)

        mapView.moveCameraFitBound(
          LatLngBounds.Builder().apply {
            for (i in 0 until coordinatesArray.size())
              include(toNaverLatLng(coordinatesArray.getMap(i)))
          }.build(),
          getInt(edgePadding, "left"),
          getInt(edgePadding, "top"),
          getInt(edgePadding, "right"),
          getInt(edgePadding, "bottom")
        )
      }
      ANIMATE_TO_COORDINATE -> mapView.setCenter(toNaverLatLng(args.getMap(0)))
      SET_LOCATION_TRACKING_MODE -> mapView.setLocationTrackingMode(args.getInt(0))
      SET_LAYER_GROUP_ENABLED -> {
        val group = args.getString(0)
        val enable = args.getBoolean(1)
        when (group) {
          NaverMap.LAYER_GROUP_BUILDING -> mapView.setLayerGroupEnabled(NaverMap.LAYER_GROUP_BUILDING, enable)
          NaverMap.LAYER_GROUP_TRANSIT -> mapView.setLayerGroupEnabled(NaverMap.LAYER_GROUP_TRANSIT, enable)
          NaverMap.LAYER_GROUP_BICYCLE -> mapView.setLayerGroupEnabled(NaverMap.LAYER_GROUP_BICYCLE, enable)
          NaverMap.LAYER_GROUP_TRAFFIC -> mapView.setLayerGroupEnabled(NaverMap.LAYER_GROUP_TRAFFIC, enable)
          NaverMap.LAYER_GROUP_CADASTRAL -> mapView.setLayerGroupEnabled(NaverMap.LAYER_GROUP_CADASTRAL, enable)
          NaverMap.LAYER_GROUP_MOUNTAIN -> mapView.setLayerGroupEnabled(NaverMap.LAYER_GROUP_MOUNTAIN, enable)
        }
      }
    }
  }

  override fun getCommandsMap(): Map<String, Int> {
    return mapOf(
      "animateToRegion" to ANIMATE_TO_REGION,
      "animateToCoordinate" to ANIMATE_TO_COORDINATE,
      "animateToTwoCoordinates" to ANIMATE_TO_TWO_COORDINATES,
      "setLocationTrackingMode" to SET_LOCATION_TRACKING_MODE,
      "animateToCoordinates" to ANIMATE_TO_COORDINATES,
      "setLayerGroupEnabled" to SET_LAYER_GROUP_ENABLED
    )
  }

  override fun getExportedCustomBubblingEventTypeConstants(): Map<String, Any>? {
    return mutableMapOf<String, Any>().apply {
      for (eventName in RNNaverMapViewProps.EVENT_NAMES)
        put(eventName, bubbled(eventName))
    }
  }

  override fun createShadowNodeInstance(): LayoutShadowNode {
    // A custom shadow node is needed in order to pass back the width/height of the map to the
    // view manager so that it can start applying camera moves with bounds.
    return SizeReportingShadowNode()
  }

  override fun addView(parent: RNNaverMapViewContainer, child: View, index: Int) {
    parent.addFeature(child, index)
    Log.e("RNN", "RNNaverMapViewManager.addView(child: ${child.javaClass.simpleName}, index: $index)")
  }

  override fun getChildCount(view: RNNaverMapViewContainer): Int {
    return view.featureCount
  }

  override fun getChildAt(view: RNNaverMapViewContainer, index: Int): View {
    return view.getFeatureAt(index)!!
  }

  override fun removeViewAt(parent: RNNaverMapViewContainer, index: Int) {
    parent.removeFeatureAt(index)
    Log.e("RNN", "RNNaverMapViewManager.removeFeatureAt(index: $index)")
  }

  override fun needsCustomLayoutForChildren(): Boolean {
    return true
  }

  private fun bubbled(callbackName: String): Map<String, Any> {
    return mapOf("phasedRegistrationNames" to mapOf("bubbled" to callbackName))
  }
}
